parser.js

```diff
@@ -24,53 +43,81 @@ function buildRuleset(name, rules) {
-    ['meta[property="twitter:title"]', node => node.element.getAttribute('content')],
-    ['meta[name="hdl"]', node => node.element.getAttribute('content')],
-    ['title', node => node.element.text],
-  ],
-  type: [
-    ['meta[property="og:type"]', node => node.element.getAttribute('content')],
-  ],
-  url: [
-    ['meta[property="og:url"]', node => node.element.getAttribute('content')],
-    ['link[rel="canonical"]', node => node.element.getAttribute('href')],
-  ],
+  description: {
+    rules: [
+      ['meta[property="og:description"]', node => node.element.getAttribute('content')],
+      ['meta[name="description"]', node => node.element.getAttribute('content')],
+    ],
+  },
+
+  icon_url: {
+    rules: [
+      ['link[rel="apple-touch-icon"]', node => node.element.getAttribute('href')],
+      ['link[rel="apple-touch-icon-precomposed"]', node => node.element.getAttribute('href')],
+      ['link[rel="icon"]', node => node.element.getAttribute('href')],
+      ['link[rel="fluid-icon"]', node => node.element.getAttribute('href')],
+      ['link[rel="shortcut icon"]', node => node.element.getAttribute('href')],
+      ['link[rel="Shortcut Icon"]', node => node.element.getAttribute('href')],
+      ['link[rel="mask-icon"]', node => node.element.getAttribute('href')],
+    ],
+    processors: [
+      (icon_url, context) => makeUrlAbsolute(context.url, icon_url)
+    ]
+  },
+
+  image_url: {
+    rules: [
+      ['meta[property="og:image:secure_url"]', node => node.element.getAttribute('content')],
+      ['meta[property="og:image:url"]', node => node.element.getAttribute('content')],
+      ['meta[property="og:image"]', node => node.element.getAttribute('content')],
+      ['meta[property="twitter:image"]', node => node.element.getAttribute('content')],
+      ['meta[name="thumbnail"]', node => node.element.getAttribute('content')],
+    ],
+    processors: [
+      (image_url, context) => makeUrlAbsolute(context.url, image_url)
+    ],
+  },
+
+  keywords: {
+    rules: [
+      ['meta[name="keywords"]', node => node.element.getAttribute('content')],
+    ],
+  },
+
+  title: {
+    rules: [
+      ['meta[property="og:title"]', node => node.element.getAttribute('content')],
+      ['meta[property="twitter:title"]', node => node.element.getAttribute('content')],
+      ['meta[name="hdl"]', node => node.element.getAttribute('content')],
+      ['title', node => node.element.text],
+    ],
+  },
+
+  type: {
+    rules: [
+      ['meta[property="og:type"]', node => node.element.getAttribute('content')],
+    ],
+  },
+
+  url: {
+    rules: [
+      ['meta[property="og:url"]', node => node.element.getAttribute('content')],
+      ['link[rel="canonical"]', node => node.element.getAttribute('href')],
+    ],
+  },
 };
 
-function getMetadata(doc, rules) {
+function getMetadata(doc, rules, context) {
   const metadata = {};
   const ruleSet = rules || metadataRules;
 
+  context = context || {};
```

Comment:
nit: I think you could also just specify a default value in the function declaration:

```js
function getMetadata(doc, rules = metadataRules, context = {}) {
  ...
}
```